incubator/hnc/pkg/validators/hierarchical_namespace.go

```diff
@@ -0,0 +1,123 @@
+package validators
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	"k8s.io/api/admission/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	api "github.com/kubernetes-sigs/multi-tenancy/incubator/hnc/api/v1alpha1"
+	"github.com/kubernetes-sigs/multi-tenancy/incubator/hnc/pkg/forest"
+	"github.com/kubernetes-sigs/multi-tenancy/incubator/hnc/pkg/reconcilers"
+)
+
+// HierarchicalNamespaceServingPath is where the validator will run. Must be kept in sync with the
+// kubebuilder markers below.
+const (
+	HierarchicalNamespaceServingPath = "/validate-hnc-x-k8s-io-v1alpha1-hierarchicalnamespaces"
+)
+
+// Note: the validating webhook FAILS CLOSE. This means that if the webhook goes down, all further
+// changes are forbidden.
+//
+// +kubebuilder:webhook:path=/validate-hnc-x-k8s-io-v1alpha1-hierarchicalnamespaces,mutating=false,failurePolicy=fail,groups="hnc.x-k8s.io",resources=hierarchicalnamespaces,verbs=create;update;delete,versions=v1alpha1,name=hierarchicalnamespaces.hnc.x-k8s.io
+
+type HierarchicalNamespace struct {
+	Log     logr.Logger
+	Forest  *forest.Forest
+	decoder *admission.Decoder
+}
+
+// req defines the aspects of the admission.Request that we care about.
+type hnsRequest struct {
+	hns *api.HierarchicalNamespace
+	op  v1beta1.Operation
+}
+
+// Handle implements the validation webhook.
+func (v *HierarchicalNamespace) Handle(ctx context.Context, req admission.Request) admission.Response {
+	log := v.Log.WithValues("Namespace", req.Namespace, "Name", req.Name)
+	// Early exit since the HNC SA can do whatever it wants.
+	if isHNCServiceAccount(&req.AdmissionRequest.UserInfo) {
+		log.V(1).Info("Allowed change by HNC SA")
+		return allow("HNC SA")
+	}
+
+	decoded, err := v.decodeRequest(log, req)
+	if err != nil {
+		v.Log.Error(err, "Couldn't decode request")
+		return deny(metav1.StatusReasonBadRequest, err.Error())
+	}
+
+	resp := v.handle(decoded)
+	log.V(1).Info("Handled", "allowed", resp.Allowed, "code", resp.Result.Code, "reason", resp.Result.Reason, "message", resp.Result.Message)
+	return resp
+}
+
+// handle implements the non-boilerplate logic of this validator, allowing it to be more easily unit
+// tested (ie without constructing a full admission.Request). It validates that the request is allowed
+// based on the current in-memory state of the forest.
+func (v *HierarchicalNamespace) handle(req *hnsRequest) admission.Response {
+	v.Forest.Lock()
+	defer v.Forest.Unlock()
+
+	pnm := req.hns.Namespace
+	cnm := req.hns.Name
+	cns := v.Forest.Get(cnm)
+
+	if req.op == v1beta1.Update {
```

Comment:
Maybe restructure this as a switch statement, with any unexpected op resulting in `log.Error`? We should let unexpected ops through though, since it's probably just a webhook misconfiguration.

I'm thinking we _should_ allow `Update` though in case people want to put labels or annotations onto them. You can't change the namespace or name, which are the only properties we care about anyway.

Reply:
As discussed above, i will remove all `update` related code since we will allow `update` requests.